### PR TITLE
Use cudnn in Convolution2D to speedup

### PR DESCRIPTION
**Pull request by entron**

### CODE REVIEW

**Comment by fchollet** [comment]
This cannot be merged since it will crash on CPU as well as on GPU on a system where cuDNN is not installed.

In theory, on GPU Theano will automatically use cuDNN if available when called from `conv2d`. In which situations is this not the case? Surely this should be addressed as a fix in Theano. If we end up specifying the use of cuDNN manually, then that should only be as a temporary workaround and that should only be done 1) after checking that the code is configured to run on GPU, and 2) after checking that cuDNN is available.

**Comment by entron** [comment]
Agree that the correct place to fix the problem is in theano rather than keras. I don't know much about the internal of theano, and it seems this problem has been there for quite some time. For the moment I just use this patch as a work around.

**Comment by fchollet** [comment]
Do you have any specific information as to *when* cuDNN is not used by `conv2d` on GPU?

**Comment by entron** [comment]
I am also curious about this. My guess is when calculating the "gradient of a strided convolution". Hope some theano expert can explain about this.

**Comment by fchollet** [comment]
What do you think about modifying this PR to check if the code is supposed to run on GPU + if cuDNN is available, and only then does a manual call to cuDNN? By default conv2d would be called.

**Comment by entron** [comment]
I added cudnn availability check, but I am not sure how to check " if the code is supposed to run on GPU".

**Comment by fchollet** [comment]
You can do this with `theano.config.device`. It has two possible string values: `"gpu"` or `"cpu"`.

**Comment by entron** [comment]
Thanks! I just added the check.

**Comment by fchollet** [comment]
Thanks! Looks good, just merged it.

Since this is essentially a workaround for a Theano issue, would you consider also raising the issue with the Theano devs?

**Comment by entron** [comment]
Ok, I will take a look.
